airgen/src/lib.rs

```diff
@@ -56,15 +63,52 @@ pub fn compile(input: AnalysisASMFile) -> PILGraph {
         instances.push((location, ty));
     }
 
+    // count incoming permutations for each machine.
+    let mut incoming_permutations = instances
+        .iter()
+        .map(|(location, _)| (location.clone(), 0))
+        .collect();
+
     // visit the tree compiling the machines
-    let objects = instances
+    let mut objects: BTreeMap<_, _> = instances
         .into_iter()
         .map(|(location, ty)| {
-            let object = ASMPILConverter::convert_machine(&location, &ty, &input);
+            let object = ASMPILConverter::convert_machine(
+                &location,
+                &ty,
+                &input,
+                &mut incoming_permutations,
+            );
             (location, object)
         })
         .collect();
 
+    // add pil code for the selector array and related constraints
+    for (location, count) in incoming_permutations {
+        let obj = objects.get_mut(&location).unwrap();
+        if count > 0 {
+            // TODO: only block machines actually need a used flag to
+            // distinguish used/unused blocks, but we don't have a way of
+            // detecting it here.
+            obj.has_used_flag = true;
+            obj.pil.extend([
+                parse_pil_statement(&format!("col witness {LINK_SELECTOR_NAME}[{count}];")),
+                parse_pil_statement(&format!(
+                    "let {LINK_USED_NAME} = std::array::sum({LINK_SELECTOR_NAME});"
+                )),
+                parse_pil_statement(&format!("std::utils::force_bool({LINK_USED_NAME});")),
+                parse_pil_statement(&format!(
+                    "std::array::map({LINK_SELECTOR_NAME}, std::utils::force_bool);"
+                )),
+            ]);
+            if let Some(latch) = &obj.latch {
+                obj.pil.push(parse_pil_statement(&format!(
+                    "std::array::map({LINK_SELECTOR_NAME}, |s| std::utils::unchanged_until(s, {latch}));"
```

Comment:
I think Poseidon is a good test case for this. We could eventually change mos of our tests to use permutations instead of lookups.

I tried implementing this suggestion and testing with the Poseidon machine. I ran into more witgen issues which are fixed here: https://github.com/powdr-labs/powdr/pull/1157

Reply:
Yep, were haven't finished the `pil` conversion at this point